examples/common.cpp

```diff
@@ -23,19 +30,37 @@ extern "C" __declspec(dllimport) int __stdcall WideCharToMultiByte(unsigned int
 #define CP_UTF8 65001
 #endif
 
-bool gpt_params_parse(int argc, char ** argv, gpt_params & params) {
-    // determine sensible default number of threads.
-    // std::thread::hardware_concurrency may not be equal to the number of cores, or may return 0.
+int32_t get_num_physical_cores() {
 #ifdef __linux__
     std::ifstream cpuinfo("/proc/cpuinfo");
-    params.n_threads = std::count(std::istream_iterator<std::string>(cpuinfo),
-                                  std::istream_iterator<std::string>(),
-                                  std::string("processor"));
-#endif
-    if (params.n_threads == 0) {
-        params.n_threads = std::max(1, (int32_t) std::thread::hardware_concurrency());
+    std::string line;
+    while (std::getline(cpuinfo, line)) {
+        if (line.find("cpu cores") != std::string::npos) {
+            line.erase(0, line.find(": ") + 2);
+            try {
+                return (int32_t) std::stoul(line);
+            } catch (std::invalid_argument& e) {} // Ignore if we could not parse
+        }
+    }
+#elif defined(__APPLE__) && defined(__MACH__)
+    int32_t num_physical_cores;
+    size_t len = sizeof(num_physical_cores);
+    int result = sysctlbyname("hw.perflevel0.physicalcpu", &num_physical_cores, &len, NULL, 0);
+    if (result == 0) {
+        return num_physical_cores;
     }
+    result = sysctlbyname("hw.physicalcpu", &num_physical_cores, &len, NULL, 0);
+    if (result == 0) {
+        return num_physical_cores;
+    }
+#elif defined(_WIN32)
+    std::cerr << "WARNING: automatic calibration not supported on Windows. Defaulting to 4 threads.\n" << std::endl;
+    return 4;
+#endif
+    return -1;
```

Comment:
If you're not going to implement the Windows core checking yourself, I wouldn't bother with a warning. I'd just make an decent attempt based on `std::thread::hardware_concurrency()` and mention that this pull request only handles Linux and macOS and someone else will have to implement it for Windows because you don't have access to a Windows build environment.

You'll want a fallback for unknown anyway so might as well make some kind of guess rather than -1.

This returns 4 if `std::thread::hardware_concurrency()` is <= 0, if there are 4 or fewer it uses all 4, if there are more it uses half.

It's not perfect, but it's not terrible either. Seems like a good rule of thumb and once WIN32 is implemented, it'll just handle edge cases. (Which, are likely the only cases to return 0 for hardware_concurrency.)

```suggestion
#elif defined(_WIN32)
    //TODO: Implement
#endif
    unsigned int n_threads = std::thread::hardware_concurrency()
    return n_threads > 0 ? (n_threads <= 4 ? n_threads : n_threads / 2) : 4;
```